### ISSUE TITLE
'firstDataRendered' event fires BEFORE data is rendered for grids having rowModelType="serverSide"

### ISSUE THREAD

**Issue by antmdvs** [ag-grid/ag-grid#2662]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[] support request => see 'Requesting Community Support'
```

**Providing a Reproducible Scenario**

https://plnkr.co/edit/5ffZIerrACtaHl82Mle4?p=preview

**Current behavior**

`onFirstDataRendered` incorrectly fires BEFORE data is rendered when `rowModelType="serverSide"`. This is contradictory to the documentation for auto-sizing columns after async data requests;
https://www.ag-grid.com/javascript-grid-resizing/#resize-after-data

which states:

> ...you can only reliably use `firstDataRendered` as the row data will be made available, and hence rendered, after the grid is ready.

Furthermore, it actually fires BEFORE `onGridReady` as shown in the browser console for the above-linked Plunk:

![image](https://user-images.githubusercontent.com/17127310/46148836-f1780f80-c236-11e8-820a-30a3e77c4aea.png)

**Expected behavior**

`onFirstDataRendered` should only fire AFTER the data is rendered as described in the docs, so that column auto-sizing APIs can work correctly.

**Please tell us about your environment:**
Windows 10

* **ag-Grid version:** X.X.X
19.0.0

* **Browser:**
* Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36

* **Language:** ES6/7

**Comment by makinggoodsoftware** [comment]
Hi,

Your ticket has been flagged as in_zendesk, it means that we recognize this as an issue worth investigating so we have moved it to our official support channel for customers (zendesk)

We will carry on there with the investigation and we will update you as soon as possible.

**Comment by makinggoodsoftware** [comment]
Thank you for your email. This is in our backlog now and we are tracking it with the following reference and description.

AG-2119 - Events - firstDataRendered fires before data is rendered with server side row model

You can follow the status of feature requests, bugs and releases using our pipeline: 

https://www.ag-grid.com/ag-grid-pipeline/
(Note that newly raised issues can take up to 2 hours to appear in the pipeline)

**Comment by snrtbnc** [comment]
I couldnt find AG-2119. Is it fixed because it seems not.

**Comment by nickharder88** [comment]
This is a problem we are having as well, I don't think it was properly logged.

**Comment by 904317681** [comment]
still having problem

**Comment by naibinh** [comment]
@makinggoodsoftware , the problem is still there.
Do you have any update?

**Comment by jjenzz** [comment]
Has anyone figured out a workaround for this that they could share? I'm still experiencing this issue with `v20.2.0`.

**Comment by davis4acca** [comment]
Hi,

This is in our backlog:  AG-3211 | Add server side event for onFirstDataLoaded

You can follow the status of feature requests, bugs and releases using our pipeline:

https://www.ag-grid.com/ag-grid-pipeline/

As a workaround this can be leveraged with getCacheBlockState api method and see if the blocks are loaded or not.

See the below plunker for an example.

https://next.plnkr.co/edit/blP2jehZwiQBczCx?preview

Regards,
Davis

**Comment by spautz** [comment]
I ran into an intermittent issue with the workaround above, where it checks `getCacheBlockState` on an interval and performs some work once loaded. Specifically, I wanted to autoSizeColumns once they were ready -- but the timing would occasionally hit a situation where the cache was loaded but the cells hadn't rendered yet, which then gave the wrong result.

This was using ag-grid 21 and 22 with React, with a few unoptimized and complex components, so cell rendering wasn't immediate.

My solution was to use ag-grid's `animationFrameService` to do the column resize only after the render had completed. It isn't perfect, but it worked for my case:
```
function autoSizeColumnsWhenReady() {
  // wait for the initial render to complete: animationQueue will be empty once it is
  if (gridApi.animationFrameService.ticking) {
    gridApi.addEventListener('animationQueueEmpty', autoSizeColumns);
  } else {
    autoSizeColumns();
  }
}

function autoSizeColumns() {
  gridApi.removeEventListener('animationQueueEmpty', autoSizeColumns);
  // waiting a single tick seems to give enough time for things to always show up
  setTimeout(() => {
    // columnController.autoSizeColumns happens in here
  }, 0);
}
```

**Comment by lwouis** [comment]
Thanks a lot @spautz for sharing your workaround! I was able to use that strategy around the animationQueue being empty, to auto-size columns on initial rendering.

I would suggest to the ag-grid team that the developer experience be improved by updating the incorrect docs, and perhaps by wrapping this workaround, or a better implementation into a user-friendly API 👍

**Comment by shvedovskiy** [comment]
@lwouis, I am trying to use this workaround in onGridReady callback for Ag Grid v24, but `gridApi.animationFrameService.ticking` is always `false`, even when I see that the loading animation in the table. Did it work for you in version 24?

**Comment by lwouis** [comment]
@shvedovskiy sorry about the confusion. I checked my code, and it's not the same as what I shared above. I don't remember why I changed it. It may have been to work in later versions of ag-grid. In any case, here is working code, for version 24:

```tsx
<AgGridReact onFirstDataRendered={onFirstDataRendered} ... />

function onFirstDataRendered(event: FirstDataRenderedEvent): void {
    autoSizeColumns(event, ['columnA', 'columnB'])
}

function autoSizeColumns(event: FirstDataRenderedEvent, columns: string[]): void {
  // auto-sizing should be done after browser rendering; we wait for the animation queue to be empty
  if (!event.api.isAnimationFrameQueueEmpty()) {
    event.api.addEventListener('animationQueueEmpty', autoSizeColumns_)
  } else {
    autoSizeColumns_(event, columns)
  }
}

function autoSizeColumns_(event: FirstDataRenderedEvent | AnimationQueueEmptyEvent, columns: string[]) {
  event.api.removeEventListener('animationQueueEmpty', autoSizeColumns_)
  event.columnApi.autoSizeColumns(columns)
}
```

**Comment by midwesterntechnology** [comment]
This is still not fixed?

**Comment by rolandarau** [comment]
Still experiencing this issue in ag-grid version **26.2.0**

**Comment by makinggoodsoftware** [comment]
Hi guys,

Note that firstDataRendered is not an event designed for the server-side row model... If you are using the SSRM, you already are always getting called into getRows, and in there, it should be fairly simple to tell if this is the first time you are getting called into

To us it seems suspicious that using the SSRM that callback is used, as with SSRM there is no such a thing as data being settled as long as the user keeps scrolling and fetching more data

Note that we raised an FR, but this is just to have a nice hook in case you need this in your app, not because this is a bug.

Note also that SSRM is an enterprise feature, what we believe is being reported here are support cases, as you are probably looking to leverage that callback to do something?

Note that if you hold an enterprise version, you are more than welcome to contact us through Zendesk and explain to us your use case, and we will gladly provide you with support and guidance

If you don't have an enterprise license, well, you should not be using this feature 😄

In any case, and for the record in case licensed users come across this ticket, if you want to do anything on the SSRM on the back of data being loaded, the correct place always to do so, is on the back of your getRows callback

Hope this clarifies things!

**Comment by cg-roling** [comment]
Looks like v28.1.1 added an event for this, but I haven't been able to figure out whether it's a new event, or if onFirstDataRendered was changed to work more like expected... anybody know for sure?

> [SSRM] Add event to fire when using server-side row model after data loaded and rendered (as onFirstDataRendered only works reliably with client-side row model and fires before data is rendered when using SSRM)

**Comment by AndrewGlazier** [comment]
Hi cg-rolling,

In version 28.1.0 the `onFirstDataRendered` event firing was changed as part of
AG-7036 - When using server-side row model the onFirstDataRendered event fires before data is loaded into the grid rows

The ticket you have referenced (AG-3211) was closed in 28.1.1 as part of our backlog housekeeping.

**Comment by cg-roling** [comment]
Perfect, thanks for the update!
